src/loader/utils/load_batch.php

```diff
@@ -28,10 +28,24 @@ function load_batch(array $config, array $transformedevents, callable $loader) {
         $loadedevents = construct_loaded_events($transformedevents, true);
         return $loadedevents;
     } catch (\Exception $e) {
+        $batchsize = count($transformedevents);
         $logerror = $config['log_error'];
-        $logerror("Failed load for event id #" . $eventobj->id . ": " .  $e->getMessage());
+        $logerror("Failed load batch (" . $batchsize . " events)" .  $e->getMessage());
         $logerror($e->getTraceAsString());
-        $loadedevents = construct_loaded_events($transformedevents, false);
+
+        // In the event of a 400 error, recursively retry sending statements in increasingly
+        // smaller batches so that only the actual bad data fails.
+        if ($batchsize === 1 || $e->getCode() !== 400 || $config['lrs_resend_failed_batches'] !== '1') {
```

Comment:
I might have flipped this if around and just returned `construct_loaded_events($transformedevents, false);` outside without an `else`, but that is very much a personal preference. I think I prefer that way because it means you don't have to reverse the condition (lacked a better phrase, hope that makes sense). You could instead write the following.

```php
$hasInvalidStatement = $e->getCode() === 400;
$isResendEnabled = $config['lrs_resend_failed_batches'] === '1';

if ($batchsize > 1 && $hasInvalidStatement && $isResendEnabled) {
  // ...
}
```

You could even extract resend logic into a function to remove the need for the comment that currently explains the if block.

```php
$hasInvalidStatement = $e->getCode() === 400;
$isResendEnabled = $config['lrs_resend_failed_batches'] === '1';

if ($batchsize > 1 && $hasInvalidStatement && $isResendEnabled) {
  retry_in_smaller_batches($config, $transformedevents, $loader));
}
```

Reply:
Yeah sounds 👍